src/server/ability.ts

```diff
@@ -0,0 +1,15 @@
+import {AbilityBuilder, Ability} from '@casl/ability'
+
+export default function defineAbilityFor(user: any) {
+  const {can, build} = new AbilityBuilder(Ability)
+
+  if (user.roles?.includes('admin')) {
+    can('manage', 'all') // read-write access to everything
+  }
+
+  if (user.roles?.includes('admin', 'editor', 'publisher')) {
+    can('upload', 'Video')
+  }
+
+  return build()
+}
```

Comment:
very similar to cancan and extremely robust in terms of how we can define permissions

https://casl.js.org/v5/en/guide/define-rules

Reply:
I refactored this to be nice and tidy. Very happy with how it turned out.